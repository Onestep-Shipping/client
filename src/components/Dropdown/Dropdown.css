.login-container {
  position: relative;
  display: flex;
  flex-direction: row-reverse;
  align-items: flex-end;
}

.user-button {
  background-color: transparent;
  color: white;
  font-size: 16px;
  border-radius: 2px;
  border: none;
  outline: none;
  cursor: pointer;
  display: flex;
  flex-direction: row;
  align-items: center;
  margin-left: 0px;
  margin-top: 5px;
  user-select: none;
}

.dropdown-menu {
  list-style-type: none;
  margin: 0;
  padding: 0;
  top: 40px;
  right: 15px;
  background-color: white;
  position: absolute;

  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);
  z-index: 1;
  margin-top: 5px;
  border: 1px solid #6d6d6d
}

.dropdown-menu-item {
  padding: 8px 16px;
  border-bottom: 1px solid #e5e5e5;
  cursor: pointer;
  line-height: 25px;
}

.dropdown-menu-item:last-child {
  border-bottom: none;
}
.dropdown-menu-item:hover {
  background-color: #e5e5e5;
  color: black;
}