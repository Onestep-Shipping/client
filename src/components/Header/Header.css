.header-container {
  width: 85%;
  height: 105px;
  border-bottom: solid 1px white;
}

.menu-container {
  width: 70%;
  height: 50%;
  display: flex;
  flex-direction: row-reverse;
  align-items: flex-end;
  justify-content: space-between;
}

.logo-container {
  width: 30%;
  height: 50%;
  display: flex;
  flex-direction: row;
  align-items: center;
  margin-bottom: 5px;
}

@media (max-width:880px) {
  .logo-container {
    margin-bottom: 50px;
  }
}

#logo-text {
  color: white;
  font-family: 'Impact';
  font-size: 50px;
  margin-bottom: 50px;
  user-select: none;
  cursor: pointer;
}

.lower-container {
  width: 100%;
  height: 50%;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  margin: 0;
  padding: 0;
}

.button-content {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
}

.svg {
	width: 30px;
	height: 30px;
  margin-right: 5px;
}

.svg-arrow {
  margin-left: 10px;
  width: 10px;
  height: 10px;
}

.svg-noti {
  width: 22px;
  height: auto;
  margin: 2.5px 2px;
 }

 .red-dot {
   width: 10px;
   height: 10px;
   background-color: red;
   border-radius: 10px;
 }
