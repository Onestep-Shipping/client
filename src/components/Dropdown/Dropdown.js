import React, {useState, useEffect, useRef, useCallback} from 'react';
import './Dropdown.css';
import PropTypes from 'prop-types';
import AuthModal from '../AuthModal/AuthModal.js';

const Dropdown = (props) => {
  const node = useRef();
  const [open, setOpen] = useState(false);
  const [isModalOpen, setModalOpen] = useState(false);

  const {type = '', content, options = [], onChange, isLoggedIn = false, isAlwaysVisible = true} = props;

  const openModal = () => {
    setModalOpen(true);
  }
 
  const closeModal = () => {
    setModalOpen(false);
  }

  const measuredRef = useCallback((e) => {
    if (node.current && !node.current.contains(e.target)) {
      setOpen(false);
    }
  }, []);

  const handleChange = selectedValue => {
    onChange(selectedValue);
    setOpen(false);
  };

  useEffect(() => {
    document.addEventListener('mousedown', measuredRef);

    return () => {
      document.removeEventListener('mousedown', measuredRef);
    };
  }, [measuredRef]);

  const handleAuth = () => {
    if (isLoggedIn) {
      setOpen(!open);
    } else {
      openModal();
    }
  }

  if (!isAlwaysVisible) {
    return <div />;
  }

  return (
    <div ref={node} className="login-container">
      <button className="user-button" onClick={handleAuth}>
        {content}
      </button>
      {open && (
        <ul className="dropdown-menu">
          {options.map((opt, ind) => (
            <li className="dropdown-menu-item" key={ind} onClick={() => handleChange(opt)}>
              {opt}
            </li>
          ))}
        </ul>
      )}
      <AuthModal
        isModalOpen={isModalOpen}
        closeModal={closeModal}
        type={type} />
    </div>
  );
};

export default Dropdown;

Dropdown.propTypes = {
  type: PropTypes.string,
  content: PropTypes.element,
  options: PropTypes.array,
  onChange: PropTypes.func,
  isLoggedIn: PropTypes.bool,
  isAlwaysVisible: PropTypes.bool
};

