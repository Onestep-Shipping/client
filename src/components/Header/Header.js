import './Header.css';

import { ADMIN_OPTIONS, USER_OPTIONS } from '../../constants/DropdownConstants.js';
import React, {useCallback, useContext} from 'react';

import { AuthContext } from '../../context/AuthContext.js';
import Dropdown from '../Dropdown/Dropdown.js';
import HeaderText from '../HeaderText/HeaderText.js';
import NOTIFICATIONS from '../../data/NotificationData.js';
import SearchTextfield from '../SearchTextfield/SearchTextfield.js';
import app from '../../firebase/base.js';
import arrowDownIcon from '../../assets/arrow-down.svg';
import loginIcon from '../../assets/login-icon.svg';
import notiIcon from '../../assets/noti-icon.svg';
import registerIcon from '../../assets/register-icon.svg';
import { useHistory } from 'react-router-dom';

const Header = () => {
  const history = useHistory();

  const { currentUser, isAdmin } = useContext(AuthContext);
  const isLoggedIn = currentUser ? true : false;

  const handleNotification = useCallback((val) => {
    console.log(val);
  }, []);

  const signOut = useCallback(() => {
    app.auth().signOut();
    history.push('/');
  }, [history]);

    const handleAdminRedirect = useCallback((val) => {
    const lcName = val.toLowerCase().replace(' ', '-');
    history.push("/admin/" + lcName);
  }, [history]);
  
  const handleLogin = useCallback((val) => {
    switch (val) {
      case "My Booking":
        history.push('/profile');
        break;
      case "Sign Out":
        signOut();
        break;
      default:
        handleAdminRedirect(val);
        break;
    }
  }, [history, signOut, handleAdminRedirect]);

  const notificationCustomStyle = {
    width: '250px',
    height: '250px', 
    overflowY: 'scroll'
  }

  const userButton = (name, icon) => {return (
    <div className="button-content" id={name === "Login" ? "add-margin" : ""}>
      <img className="svg" src={icon} alt="User Icon" />
      <span>
      {currentUser ?
        currentUser.email.substring(0, currentUser.email.indexOf('@')) : name}
      </span>
      {currentUser && <img className="svg-arrow" src={arrowDownIcon} alt="Dropdown Icon" />}
    </div>
  )};

  const notificationButton = (
    <div className="button-content">
      <img className="svg-noti" src={notiIcon} alt="Nofification Icon" />
      {NOTIFICATIONS.length > 0 && <div className="red-dot" />}
    </div>
  );

  return (
    <div className="header-container">
      <div onClick={() => history.push('/')}>
        <span className="logo-text">OneStep Shipping</span>
      </div>

      <div className="menu-container">
        <div className="login-container">
          <Dropdown
            content={userButton('Login', loginIcon)}
            type={'login'}
            onChange={val => handleLogin(val)}
            options={isAdmin ? ADMIN_OPTIONS : USER_OPTIONS}
            isLoggedIn={isLoggedIn} />
          {!isLoggedIn &&
          <Dropdown
            type={'register'}
            content={userButton('Register', registerIcon)}
            onChange={val => handleLogin(val)}
            isLoggedIn={isLoggedIn} 
          />}
          {isLoggedIn &&
          <Dropdown
            type={'noti'}
            content={notificationButton}
            onChange={val => handleNotification(val)}
            options={NOTIFICATIONS.length > 0 ? NOTIFICATIONS : ["Notification is empty."]}
            isLoggedIn={isLoggedIn}
            customStyle={notificationCustomStyle} 
          />}
        </div>
        <SearchTextfield />
        <HeaderText value="Contact" action={() => history.push('/contact')} />
        <HeaderText value="News" action={() => history.push('/news')} />
        <HeaderText value="Services" action={() => history.push('/services')} />
        <HeaderText value="About" action={() => history.push('/about')} />
      </div>
    </div>
  );
};

export default Header;
