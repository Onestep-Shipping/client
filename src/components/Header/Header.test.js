import AuthProvider from '../../context/AuthContext.js';
import Dropdown from '../Dropdown/Dropdown.js';
import Header from './Header';
import React from 'react';
import { mount } from 'enzyme';

const wrapper = mount(
  <AuthProvider>
    <Header />
  </AuthProvider>
);

const logoText = "OneStep Shipping";

describe('Header', () => {
  it('should render the Header Component correctly', () => {   
    expect(wrapper).toMatchSnapshot();
  });
  it('should renders the Logo text', () => { 
    expect(wrapper.find('.logo-text').text()).toEqual(logoText);
  });
  it('should alway have only 2 dropdowns', () => {
    expect(wrapper.find(Dropdown).length).toBe(2); 
  });
});