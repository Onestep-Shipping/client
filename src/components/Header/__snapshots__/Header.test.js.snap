// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render the Header Component correctly 1`] = `
<AuthProvider>
  <Header>
    <div
      className="header-container"
    >
      <div
        onClick={[Function]}
      >
        <span
          className="logo-text"
        >
          OneStep Shipping
        </span>
      </div>
      <div
        className="menu-container"
      >
        <div
          className="login-container"
        >
          <Dropdown
            content={
              <div
                className="button-content"
                id="add-margin"
              >
                <img
                  alt="User Icon"
                  className="svg"
                  src="login-icon.svg"
                />
                <span>
                  Login
                </span>
              </div>
            }
            isLoggedIn={false}
            onChange={[Function]}
            options={
              Array [
                "My Booking",
                "Sign Out",
              ]
            }
            type="login"
          >
            <div
              className="login-container"
            >
              <button
                className="user-button"
                onClick={[Function]}
              >
                <div
                  className="button-content"
                  id="add-margin"
                >
                  <img
                    alt="User Icon"
                    className="svg"
                    src="login-icon.svg"
                  />
                  <span>
                    Login
                  </span>
                </div>
              </button>
              <AuthModal
                closeModal={[Function]}
                isModalOpen={false}
                type="login"
              >
                <Modal
                  ariaHideApp={true}
                  bodyOpenClassName="ReactModal__Body--open"
                  closeTimeoutMS={0}
                  isOpen={false}
                  onRequestClose={[Function]}
                  parentSelector={[Function]}
                  portalClassName="ReactModalPortal"
                  role="dialog"
                  shouldCloseOnEsc={true}
                  shouldCloseOnOverlayClick={true}
                  shouldFocusAfterRender={true}
                  shouldReturnFocusAfterClose={true}
                  style={
                    Object {
                      "content": Object {
                        "border": "none",
                        "borderRadius": "10px",
                        "bottom": "auto",
                        "left": "50%",
                        "padding": "0px",
                        "right": "auto",
                        "top": "50%",
                        "transform": "translate(-50%, -50%)",
                      },
                    }
                  }
                >
                  <Portal
                    containerInfo={
                      <div
                        class="ReactModalPortal"
                      />
                    }
                  >
                    <ModalPortal
                      ariaHideApp={true}
                      bodyOpenClassName="ReactModal__Body--open"
                      closeTimeoutMS={0}
                      defaultStyles={
                        Object {
                          "content": Object {
                            "WebkitOverflowScrolling": "touch",
                            "background": "#fff",
                            "border": "1px solid #ccc",
                            "borderRadius": "4px",
                            "bottom": "40px",
                            "left": "40px",
                            "outline": "none",
                            "overflow": "auto",
                            "padding": "20px",
                            "position": "absolute",
                            "right": "40px",
                            "top": "40px",
                          },
                          "overlay": Object {
                            "backgroundColor": "rgba(255, 255, 255, 0.75)",
                            "bottom": 0,
                            "left": 0,
                            "position": "fixed",
                            "right": 0,
                            "top": 0,
                          },
                        }
                      }
                      isOpen={false}
                      onRequestClose={[Function]}
                      parentSelector={[Function]}
                      portalClassName="ReactModalPortal"
                      role="dialog"
                      shouldCloseOnEsc={true}
                      shouldCloseOnOverlayClick={true}
                      shouldFocusAfterRender={true}
                      shouldReturnFocusAfterClose={true}
                      style={
                        Object {
                          "content": Object {
                            "border": "none",
                            "borderRadius": "10px",
                            "bottom": "auto",
                            "left": "50%",
                            "padding": "0px",
                            "right": "auto",
                            "top": "50%",
                            "transform": "translate(-50%, -50%)",
                          },
                        }
                      }
                    />
                  </Portal>
                </Modal>
              </AuthModal>
            </div>
          </Dropdown>
          <Dropdown
            content={
              <div
                className="button-content"
                id=""
              >
                <img
                  alt="User Icon"
                  className="svg"
                  src="register-icon.svg"
                />
                <span>
                  Register
                </span>
              </div>
            }
            isLoggedIn={false}
            onChange={[Function]}
            type="register"
          >
            <div
              className="login-container"
            >
              <button
                className="user-button"
                onClick={[Function]}
              >
                <div
                  className="button-content"
                  id=""
                >
                  <img
                    alt="User Icon"
                    className="svg"
                    src="register-icon.svg"
                  />
                  <span>
                    Register
                  </span>
                </div>
              </button>
              <AuthModal
                closeModal={[Function]}
                isModalOpen={false}
                type="register"
              >
                <Modal
                  ariaHideApp={true}
                  bodyOpenClassName="ReactModal__Body--open"
                  closeTimeoutMS={0}
                  isOpen={false}
                  onRequestClose={[Function]}
                  parentSelector={[Function]}
                  portalClassName="ReactModalPortal"
                  role="dialog"
                  shouldCloseOnEsc={true}
                  shouldCloseOnOverlayClick={true}
                  shouldFocusAfterRender={true}
                  shouldReturnFocusAfterClose={true}
                  style={
                    Object {
                      "content": Object {
                        "border": "none",
                        "borderRadius": "10px",
                        "bottom": "auto",
                        "left": "50%",
                        "padding": "0px",
                        "right": "auto",
                        "top": "50%",
                        "transform": "translate(-50%, -50%)",
                      },
                    }
                  }
                >
                  <Portal
                    containerInfo={
                      <div
                        class="ReactModalPortal"
                      />
                    }
                  >
                    <ModalPortal
                      ariaHideApp={true}
                      bodyOpenClassName="ReactModal__Body--open"
                      closeTimeoutMS={0}
                      defaultStyles={
                        Object {
                          "content": Object {
                            "WebkitOverflowScrolling": "touch",
                            "background": "#fff",
                            "border": "1px solid #ccc",
                            "borderRadius": "4px",
                            "bottom": "40px",
                            "left": "40px",
                            "outline": "none",
                            "overflow": "auto",
                            "padding": "20px",
                            "position": "absolute",
                            "right": "40px",
                            "top": "40px",
                          },
                          "overlay": Object {
                            "backgroundColor": "rgba(255, 255, 255, 0.75)",
                            "bottom": 0,
                            "left": 0,
                            "position": "fixed",
                            "right": 0,
                            "top": 0,
                          },
                        }
                      }
                      isOpen={false}
                      onRequestClose={[Function]}
                      parentSelector={[Function]}
                      portalClassName="ReactModalPortal"
                      role="dialog"
                      shouldCloseOnEsc={true}
                      shouldCloseOnOverlayClick={true}
                      shouldFocusAfterRender={true}
                      shouldReturnFocusAfterClose={true}
                      style={
                        Object {
                          "content": Object {
                            "border": "none",
                            "borderRadius": "10px",
                            "bottom": "auto",
                            "left": "50%",
                            "padding": "0px",
                            "right": "auto",
                            "top": "50%",
                            "transform": "translate(-50%, -50%)",
                          },
                        }
                      }
                    />
                  </Portal>
                </Modal>
              </AuthModal>
            </div>
          </Dropdown>
        </div>
        <SearchTextfield>
          <div
            className="search-container"
          >
            <input
              id="search"
              placeholder="Search"
              type="text"
            />
            <button
              id="search-button"
            >
              <img
                alt="Search Icon"
                id="search-icon"
                src="search-icon.png"
              />
            </button>
          </div>
        </SearchTextfield>
        <HeaderText
          action={[Function]}
          value="Contact"
        >
          <div
            className="header-text-container"
            onClick={[Function]}
          >
            <span
              className="header-text"
            >
              Contact
            </span>
          </div>
        </HeaderText>
        <HeaderText
          action={[Function]}
          value="News"
        >
          <div
            className="header-text-container"
            onClick={[Function]}
          >
            <span
              className="header-text"
            >
              News
            </span>
          </div>
        </HeaderText>
        <HeaderText
          action={[Function]}
          value="Services"
        >
          <div
            className="header-text-container"
            onClick={[Function]}
          >
            <span
              className="header-text"
            >
              Services
            </span>
          </div>
        </HeaderText>
        <HeaderText
          action={[Function]}
          value="About"
        >
          <div
            className="header-text-container"
            onClick={[Function]}
          >
            <span
              className="header-text"
            >
              About
            </span>
          </div>
        </HeaderText>
      </div>
    </div>
  </Header>
</AuthProvider>
`;
